### PR TITLE
80 attachment consumer

### PR DESCRIPTION
**Pull request by janid1967**
https://github.com/equinor/cc-toolbox/issues/80

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Infrastructure/Migrations/20240611051747_ProCoSys4LastUpdated.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                type: "datetime2",
+                nullable: false,
+                defaultValue: new DateTime(1, 1, 1, 0, 0, 0, 0, DateTimeKind.Unspecified))
+                .Annotation("SqlServer:IsTemporal", true)
+                .Annotation("SqlServer:TemporalHistoryTableName", "AttachmentsHistory")
+                .Annotation("SqlServer:TemporalHistoryTableSchema", null)
+                .Annotation("SqlServer:TemporalPeriodEndColumnName", "PeriodEnd")
+                .Annotation("SqlServer:TemporalPeriodStartColumnName", "PeriodStart");
+
+            migrationBuilder.UpdateData(
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Infrastructure/Migrations/20240611051747_ProCoSys4LastUpdated.cs:62]: @TordJoranger There is something wrong with your latest migration where you seed the Unknown-item
When we now add a new migration without any changes, this migrationBuilder.UpdateData is made by EF

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/AttachmentAggregate/Attachment.cs [hunk 2]
```diff
@@ -54,6 +54,8 @@ public IOrderedEnumerable<Label> GetOrderedNonVoidedLabels()
     public Person? ModifiedBy { get; private set; }
     public Guid Guid { get; private set; }
     public int RevisionNumber { get; private set; }
+    public DateTime ProCoSys4LastUpdated { get; set; }
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/AttachmentAggregate/Attachment.cs:57]: I don't think we want this for non oracle mastered entities

Comment (TordJoranger) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/AttachmentAggregate/Attachment.cs:57]: after discussion with JI.
we need this, will have to add to PunchItem also

File: src/Equinor.ProCoSys.Completion.Infrastructure/Repositories/PersonRepository.cs [hunk 3]
```diff
@@ -47,13 +47,23 @@ public async Task<List<Person>> GetOrCreateManyAsync(IEnumerable<Guid> oids, Can
 
     public async Task<Person> GetOrCreateAsync(Guid oid, CancellationToken cancellationToken)
     {
+        // don't bother to check neither local db nor cache for person without real oid
+        if (oid == Guid.Empty)
+        {
+            throw new Exception($"Illegal to get or create person with null oid {oid}");
```
[the remark this answers is not in the document]
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Infrastructure/Repositories/PersonRepository.cs:53]: I'll fix this one in other branch ... I close this PR without merge due to migrations

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemAttachmentEventConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        {
+            throw new Exception($"{nameof(PunchItemAttachmentEvent)} is missing {nameof(PunchItemAttachmentEvent.Uri)}");
+        }
+    }
+
+    private void MapFromEventToAttachment(PunchItemAttachmentEvent busEvent, Attachment attachment)
+    {
+        attachment.ProCoSys4LastUpdated = busEvent.LastUpdated;
+        attachment.SyncTimestamp = DateTime.UtcNow;
+        if (!string.IsNullOrEmpty(busEvent.Title))
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemAttachmentEventConsumer.cs:179]: Why this check? 
Can title be null or empty? and if it is, does that not mean that its been cleared in oracle?

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemAttachmentEventConsumer.cs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        else
+        {
+            ValidateAttachmentMessageAsCompletionLink(busEvent);
+            if (!await HandleLinkEvent(context, busEvent))
+            {
+                return;
+            }
+        }
+
+        await unitOfWork.SaveChangesAsync(context.CancellationToken);
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemAttachmentEventConsumer.cs:48]: need to use the sync that does not set modified/created (SaveChangesFromSyncAsync)
We also need to set modifiedBy/At mannualy in syncProperties.
Also for updates

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemAttachmentEventConsumer.cs:48]: nice catch!

**Pull request closed without merging**